examples/gnrc_border_router/main.c

```diff
@@ -26,11 +26,134 @@
+#include "net/gnrc/ipv6/netif.h"
+#include "net/gnrc/netapi.h"
+#include "net/gnrc/netif.h"
+#include "net/ipv6/addr.h"
+#include "net/netdev2.h"
+#include "net/netopt.h"
+
+#include "uhcp.h"
+#include "fmt.h"
+
+kernel_pid_t gnrc_border_interface;
+kernel_pid_t gnrc_wireless_interface;
+
+static void set_interface_roles(void)
+{
+    kernel_pid_t ifs[GNRC_NETIF_NUMOF];
+    size_t numof = gnrc_netif_get(ifs);
+    int res;
+    uint16_t dev_type;
+
+    for (size_t i = 0; i < numof && i < GNRC_NETIF_NUMOF; i++) {
+        kernel_pid_t dev = ifs[i];
+        res = gnrc_netapi_get(dev, NETOPT_DEVICE_TYPE, 0, &dev_type, sizeof(dev_type));
+        if (res <= 0) {
+            dev_type = NETDEV2_TYPE_UNKNOWN;
+        }
+        if ((!gnrc_border_interface) && (dev_type == NETDEV2_TYPE_ETHERNET)) {
+            ipv6_addr_t addr, defroute;
+            gnrc_border_interface = dev;
+
+            ipv6_addr_from_str(&addr, "fe80::2");
+            gnrc_ipv6_netif_add_addr(dev, &addr, 64,
+                                     GNRC_IPV6_NETIF_ADDR_FLAGS_UNICAST);
+
+            ipv6_addr_from_str(&defroute, "::");
+            ipv6_addr_from_str(&addr, "fe80::1");
+            fib_add_entry(&gnrc_ipv6_fib_table, dev, defroute.u8, 16,
+                    FIB_FLAG_NET_PREFIX, addr.u8, 16, 0,
+                    (uint32_t)FIB_LIFETIME_NO_EXPIRE);
+        }
+        else if ((!gnrc_wireless_interface) && (dev_type == NETDEV2_TYPE_UNKNOWN)) {
+            gnrc_wireless_interface = dev;
+        }
+
+        if (gnrc_border_interface && gnrc_wireless_interface) {
+            break;
+        }
+    }
+
+    printf("Using %u as border interface and %u as wireless interface.\n", gnrc_border_interface, gnrc_wireless_interface);
+}
+
+static ipv6_addr_t _prefix;
+
+void uhcp_handle_prefix(uint8_t *prefix, uint8_t prefix_len, uint16_t lifetime, uint8_t *src, uhcp_iface_t iface)
+{
+    eui64_t iid;
+    if (!gnrc_wireless_interface) {
+        puts("uhcp_handle_prefix(): received prefix, but don't know any wireless interface");
+        return;
+    }
+
+    if (iface != gnrc_border_interface) {
+        puts("uhcp_handle_prefix(): received prefix from unexpected interface");
+        return;
+    }
+
+    if (gnrc_netapi_get(gnrc_wireless_interface, NETOPT_IPV6_IID, 0, &iid,
+                        sizeof(eui64_t)) >= 0) {
+        ipv6_addr_set_aiid((ipv6_addr_t*)prefix, iid.uint8);
+    }
+    else {
+        puts("uhcp_handle_prefix(): cannot get IID of wireless interface");
+        return;
+    }
+
+    if (ipv6_addr_equal(&_prefix, (ipv6_addr_t*)prefix)) {
+        puts("uhcp_handle_prefix(): got same prefix again");
+        return;
+    }
+
+    gnrc_ipv6_netif_add_addr(gnrc_wireless_interface, (ipv6_addr_t*)prefix, 64,
+                             GNRC_IPV6_NETIF_ADDR_FLAGS_UNICAST |
+                             GNRC_IPV6_NETIF_ADDR_FLAGS_NDP_AUTO);
+
+    gnrc_ipv6_netif_remove_addr(gnrc_wireless_interface, &_prefix);
```

Comment:
@authmillenon I agree with @kaspar030's implementation here, we want to only keep one UHCP-assigned prefix, so that if we restart the PC side we will remove the old prefix and add the new one.